[CoreLib] Iterable.withIterator

drxddy opened the pull request:
Introduced a new factory constructor for `Iterable` class  `withIterator` 

Closes #55961 

### Usage
```dart
final numbers = Iterable.withIterator(() => [1, 2, 3].iterator);
print(numbers.toList()); // [1, 2, 3]
```

```dart
final numbers = [1, 2, 3];
final iterator = numbers.iterator;
iterator.moveNext();
var lazyIterable = Iterable.withIterator(() => iterator);

// Test that the first element is already consumed
var elements = lazyIterable.toList();  // [2, 3]
Expect.isNotEmpty(elements);
Expect.equals(2, elements.first);
```

drxddy committed:
[CoreLib] Iterable.fromIteratorFactory
Introduced a new factory constructor fromIteratorFactory to create an iterable from iterator

drxddy committed:
chore: rename to withIterable

drxddy commented:
@lrhn renamed to `Iterable.withIterator` in this change [fa07cf8](https://github.com/dart-lang/sdk/pull/59908/commits/fa07cf80a24a8bc2d59024d3521f157b33260ba4), let me know if anything needs to be changed

drxddy committed:
fix: Iterable.withIterator docs and consolidated tests

drxddy committed:
Updated CHANGELOG.md: Added Iterable.withIterator

drxddy committed:
Merge branch 'main' into feat/iterable-from-iterator-factory

drxddy committed:
fix: test file rename to iterable_with_iterator_test

drxddy commented:
Hey, I've resolved all the changes suggested by @lrhn, please let me if anything can be better before we can merge it with main

lrhn commented on sdk/lib/core/iterable.dart:
```diff
@@ -915,6 +933,17 @@ class _GeneratorIterable<E> extends ListIterable<E> {
   static int _id(int n) => n;
 }
 
+class _FromIteratorIterable<E> extends Iterable<E> {
```
Could rename it `withIteratorIterable` to match the constructor name. 
(It's private, so nobody will know.)

lrhn replied:
Maybe add doc saying that it's used by `Iterable.withIteratot`.

lrhn commented on sdk/lib/core/iterable.dart:
```diff
@@ -915,6 +933,17 @@ class _GeneratorIterable<E> extends ListIterable<E> {
+
+  /// Creates an iterable that gets its elements from iterators created by
+  /// the provided factory function.
+  _FromIteratorIterable(Iterator<E> Function() iteratorFactory)
```
It's OK to just use `this._iteratorFactory` when the class is private.

lrhn reviewed (approved):
LGTM. Small suggestions on the private class, nothing visible.

drxddy committed:
refactor: renaming private class and added doc comments

drxddy commented:
Hi @lrhn, I've made these changes the latest commit [458904f](https://github.com/dart-lang/sdk/pull/59908/commits/458904f45aa9a3d5c3c944a2c849d690096013b9)
- renamed the private class to `_WithIteratorIterable`
- added doc comments explaining `_WithIteratorIterable` usage
- changed parameter construction to `this._iteratorFactory`

Verified the SDK compilation on my local machine
<img width="822" alt="Screenshot 2025-01-17 at 1 21 38 PM" src="https://github.com/user-attachments/assets/74d56d26-9cfa-49b1-8e62-e593bab85a3a" />

Verified the newly added test was passing
<img width="987" alt="Screenshot 2025-01-17 at 1 24 37 PM" src="https://github.com/user-attachments/assets/c82f4c57-4b12-401f-9a56-56b469b1ae7e" />

Verified by running an example using the locally compiled SDK
<img width="689" alt="Screenshot 2025-01-17 at 1 26 41 PM" src="https://github.com/user-attachments/assets/7bbf1bef-ae57-4a9d-b8df-22eb2da914a7" />

drxddy committed:
fix: docs white space and shorten description

drxddy committed:
fmt: dart format iterable_with_iterator_test.dart

drxddy commented:
Hey @lrhn, I did `dart format` for the newly added test file to pass the pre-submit try test, I couldn't restart that failed test myself (access restricted), I'm curious to know when you added the CL to commit-queue, does it automatically kick-off all the relevant tests?

lrhn commented:
Yes, all I did was press the "commit" button. There is also a button to run tests without committing. The commit button only runs tests of the same patch hasn't already run the tests successfully.

drxddy commented:
Got it @lrhn, are there any changes I need to make or can we merge it?

drxddy commented:
Hey @lrhn, are there any changes required to merge this?
